### ISSUE TITLE
Increase timeout when using data_dir option

### ISSUE THREAD

**Issue by anadahz** [TheTorProject/ooni-probe#543]
We should increase the timeout when data_dir option is used in ooniprobe.

When `data_dir` is not defined:

```
# time ooniprobe -f /etc/ooniprobe/ooniprobe.conf -v blocking/http_requests --url http://ooni.torproject.org
WARNING: running ooniprobe involves some risk that varies greatly
         from country to country. You should be aware of this when
         running the tool. Read more about this in the manpage or README.
Log opened.
[D] No test deck detected
Starting Tor...
[D] Setting control port as 37881
[D] Setting SOCKS port as 11335
Spawning tor process with DataDirectory /tmp/tortmpAXX2u6
5%: Connecting to directory server
10%: Finishing handshake with directory server
15%: Establishing an encrypted directory connection
20%: Asking for networkstatus consensus
25%: Loading networkstatus consensus
40%: Loading authority key certs
45%: Asking for relay descriptors
45%: Asking for relay descriptors
50%: Loading relay descriptors
51%: Loading relay descriptors
52%: Loading relay descriptors
52%: Loading relay descriptors
52%: Loading relay descriptors
53%: Loading relay descriptors
56%: Loading relay descriptors
58%: Loading relay descriptors
58%: Loading relay descriptors
62%: Loading relay descriptors
67%: Loading relay descriptors
74%: Loading relay descriptors
80%: Connecting to the Tor network
90%: Establishing a Tor circuit
100%: Done
[D] Building a TorState
Successfully bootstrapped Tor
[D] We now have the following circuits:
[...]
Found your IP via Tor
Fetching required net test inputs...
[D] Fetching and verifying inputs
Looking up collector and test helpers
[D] Got this backend error message {u'error': 404}
Setting collector and test helpers for http_requests
Creating report with OONIB Reporter. Please be patient.
This may take up to 1-2 minutes...
[...]
[D] Successfully performed report <ooni.tasks.ReportEntry object at 0x756b2a30>
[D] None
Main loop terminated.

real    9m23.158s
user    0m24.010s
sys     0m1.080s
```

With `data_dir: /opt/ooni/tor_data_dir`

```
# time ooniprobe -f /etc/ooniprobe/ooniprobe.conf -v blocking/http_requests --url http://ooni.torproject.org
WARNING: running ooniprobe involves some risk that varies greatly
         from country to country. You should be aware of this when
         running the tool. Read more about this in the manpage or README.
Log opened.
[D] No test deck detected
Starting Tor...
[D] Setting control port as 25574
[D] Setting SOCKS port as 18973
Spawning tor process with DataDirectory /opt/ooni/tor_data_dir
Unhandled Error
Traceback (most recent call last):
Failure: exceptions.RuntimeError: Tor exited with error-code 0

Traceback (most recent call last):
Failure: exceptions.RuntimeError: Tor exited with error-code 0
[!] Could not start the director
[!] An unhandled exception occurred while starting the director!
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 434, in errback
    self._startRunCallbacks(fail)
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 1184, in gotResult
    _inlineCallbacks(r, g, deferred)
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-1.5.1-py2.7.egg/ooni/director.py", line 134, in start
    yield self.startTor()
  File "/usr/local/lib/python2.7/dist-packages/Twisted-16.2.0-py2.7-linux-armv7l.egg/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-1.5.1-py2.7.egg/ooni/director.py", line 331, in setup_failed
    raise errors.UnableToStartTor
ooni.errors.UnableToStartTor: 
Main loop terminated.

real    0m24.755s
user    0m18.360s
sys     0m1.180s

```

**Comment by hellais** [comment]
I assume you mean **decrease** the timeout when the data_dir is used? If the data-dir option is used the consensus documents may already be cached and hence it will take less to bootstrap tor.

Is this what you mean?

Though given the fact that we don't know if they are cached or not until we start Tor I think the safest thing is to have the timeout value be some upper bound.

 I am not sure if it's wise to do that as in the general case the timeout will never be reached.

What is the specific issue that you are experiencing? Are you worried about the fact that it will take too long for ooniprobe to bail when it's unable to establish a Tor connection?

**Comment by anadahz** [comment]
No I actually meant *increase* the timeout time when `data_dir` option is enabled.
In the [log report](https://github.com/TheTorProject/ooni-probe/issues/543#issue-160421007) I wasn't able to perform the test `blocking/http_requests --url http://ooni.torproject.org` while without the `data_dir` option the test was successful, thought it took more than 9 minutes to complete which was actually due to tor bootstrapping that slow. 
In the contrary when the `data_dir` option was specified the test failed (due to tor not bootstrapping) after 24 seconds.

**Comment by hellais** [comment]
I don't think that what you are encountering there is a timeout being triggered, but is more likely to be an issue with Tor bootstrapping. What I think is most likely to be happening there is that there are some permissions issues with the `data_dir` and Tor is bailing immediately.

Currently there are no timeouts on the bootstrapping of Tor and when you specify a `data_dir` it should actually take **less** time on average not more.

**Comment by hellais** [comment]
Not sure what should be done about this issue. If you believe there should be some resolution please re-open it with a description of what you would like changed.

**Comment by anadahz** [comment]
@hellais I believe that is quite clear from the title and description of the title to increase the timeout when the `data_dir` option is being used, since the `data_dir` has no timeout  set then we should perhaps investigate other reasons.
Were you able to reproduce this issue?

This issue has nothing to do with permissions. When there is an issue with permission of the `data_dir` you normally get an error message similar to: `exceptions.OSError: [Errno 13] Permission denied`

**Comment by hellais** [comment]
As I stated above the problem here is that Tor is exiting with a non zero exit code. OSErrors as the ones you suggest above are triggered when **python** triggers a permissions problem, but this error is the `tor` executable failing (most probably due to a permissions issue with the `data_dir`).

I can reproduce this when I set a `data_dir` that has wrong permissions. Increasing the timeout will not only not resolve this issue, but also should not be done since in theory when you have a data_dir the time Tor takes to bootstrap should be **less** not more.

Have you tried setting correct permissions on the `data_dir` (on debian it needs to be 700 and owned by user debian-tor)?
